File: mobile/android/ASLRecognizer.java
package com.example.aslrecognizer;

import android.content.Context;
import android.graphics.Bitmap;
import android.util.Log;

import org.tensorflow.lite.Interpreter;
import org.tensorflow.lite.support.common.FileUtil;
import org.tensorflow.lite.support.image.TensorImage;
import org.tensorflow.lite.support.tensorbuffer.TensorBuffer;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * 🤖 OpenHands ASL Recognition for Android
 * TensorFlow Lite implementation with pose-based inference
 */
public class ASLRecognizer {
    private static final String TAG = "ASLRecognizer";
    
    // Model configuration
    private static final String MODEL_PATH = "asl_model_int8.tflite";
    private static final int INPUT_SIZE = 158;  // Pose features (hand + body)
    private static final int OUTPUT_SIZE = 2000; // ASL vocabulary
    private static final float CONFIDENCE_THRESHOLD = 0.7f;
    
    // TensorFlow Lite components
    private Interpreter interpreter;
    private String[] labels;
    private Context context;
    
    // Pose extractor
    private PoseExtractor poseExtractor;
    
    public ASLRecognizer(Context context) throws IOException {
        this.context = context;
        initializeModel();
        initializePoseExtractor();
        loadLabels();
    }
    
    /**
     * Initialize TensorFlow Lite model
     */
    private void initializeModel() throws IOException {
        Log.d(TAG, "🚀 Initializing OpenHands ASL model...");
        
        try {
            // Load quantized model
            ByteBuffer modelBuffer = FileUtil.loadMappedFile(context, MODEL_PATH);
            
            // Configure interpreter options
            Interpreter.Options options = new Interpreter.Options();
            options.setNumThreads(4); // Optimize for mobile CPU
            options.setUseNNAPI(true); // Use Android Neural Networks API
            
            // Create interpreter
            interpreter = new Interpreter(modelBuffer, options);
            
            Log.d(TAG, "✅ Model loaded successfully");
            Log.d(TAG, "📊 Input shape: " + java.util.Arrays.toString(interpreter.getInputTensor(0).shape()));
            Log.d(TAG, "📊 Output shape: " + java.util.Arrays.toString(interpreter.getOutputTensor(0).shape()));
            
        } catch (IOException e) {
            Log.e(TAG, "❌ Failed to load model: " + e.getMessage());
            throw e;
        }
    }
    
    /**
     * Initialize pose extractor (MediaPipe equivalent)
     */
    private void initializePoseExtractor() {
        Log.d(TAG, "🖐️ Initializing pose extractor...");
        poseExtractor = new PoseExtractor(context);
        Log.d(TAG, "✅ Pose extractor ready");
    }
    
    /**
     * Load ASL vocabulary labels
     */
    private void loadLabels() throws IOException {
        Log.d(TAG, "📚 Loading ASL vocabulary...");
        
        try {
            List<String> labelList = FileUtil.loadLabels(context, "asl_labels.txt");
            labels = labelList.toArray(new String[0]);
            
            Log.d(TAG, "✅ Loaded " + labels.length + " ASL signs");
            
        } catch (IOException e) {
            Log.e(TAG, "❌ Failed to load labels: " + e.getMessage());
            // Create dummy labels as fallback
            labels = new String[OUTPUT_SIZE];
            for (int i = 0; i < OUTPUT_SIZE; i++) {
                labels[i] = "SIGN_" + i;
            }
            Log.w(TAG, "⚠️ Using dummy labels");
        }
    }
    
    /**
     * Recognize ASL sign from camera frame
     * 
     * @param bitmap Camera frame as bitmap
     * @return Recognition result
     */
    public ASLResult recognizeSign(Bitmap bitmap) {
        long startTime = System.currentTimeMillis();
        
        try {
            // Extract pose features
            float[] poseFeatures = poseExtractor.extractPoseFeatures(bitmap);
            
            if (poseFeatures == null) {
                return new ASLResult("", 0.0f, "No hands detected", 
                                   System.currentTimeMillis() - startTime);
            }
            
            // Prepare input tensor
            ByteBuffer inputBuffer = ByteBuffer.allocateDirect(INPUT_SIZE * 4);
            inputBuffer.order(ByteOrder.nativeOrder());
            
            for (float feature : poseFeatures) {
                inputBuffer.putFloat(feature);
            }
            inputBuffer.rewind();
            
            // Prepare output tensor
            ByteBuffer outputBuffer = ByteBuffer.allocateDirect(OUTPUT_SIZE * 4);
            outputBuffer.order(ByteOrder.nativeOrder());
            
            // Run inference
            interpreter.run(inputBuffer, outputBuffer);
            
            // Process output
            outputBuffer.rewind();
            float[] probabilities = new float[OUTPUT_SIZE];
            for (int i = 0; i < OUTPUT_SIZE; i++) {
                probabilities[i] = outputBuffer.getFloat();
            }
            
            // Find best prediction
            int bestIndex = 0;
            float bestConfidence = probabilities[0];
            
            for (int i = 1; i < probabilities.length; i++) {
                if (probabilities[i] > bestConfidence) {
                    bestConfidence = probabilities[i];
                    bestIndex = i;
                }
            }
            
            // Create result
            String predictedSign = bestIndex < labels.length ? labels[bestIndex] : "UNKNOWN";
            String message = bestConfidence > CONFIDENCE_THRESHOLD ? 
                           "Recognized: " + predictedSign : 
                           "Low confidence: " + String.format("%.2f", bestConfidence);
            
            return new ASLResult(
                bestConfidence > CONFIDENCE_THRESHOLD ? predictedSign : "",
                bestConfidence,
                message,
                System.currentTimeMillis() - startTime
            );
            
        } catch (Exception e) {
            Log.e(TAG, "❌ Recognition error: " + e.getMessage());
            return new ASLResult("", 0.0f, "Error: " + e.getMessage(),
                               System.currentTimeMillis() - startTime);
        }
    }
    
    /**
     * Get top N predictions
     */
    public List<ASLResult> getTopPredictions(Bitmap bitmap, int topN) {
        // Implementation similar to recognizeSign but returns multiple results
        List<ASLResult> results = new ArrayList<>();
        // ... (implementation details)
        return results;
    }
    
    /**
     * Build sentence from sequence of recognitions
     */
    public String buildSentence(List<ASLResult> recognitionHistory) {
        if (recognitionHistory.isEmpty()) {
            return "";
        }
        
        List<String> sentenceParts = new ArrayList<>();
        String currentSign = null;
        
        // Group consecutive similar signs
        for (ASLResult result : recognitionHistory) {
            if (result.getConfidence() > CONFIDENCE_THRESHOLD && 
                !result.getPredictedSign().isEmpty() &&
                !result.getPredictedSign().equals(currentSign)) {
                
                sentenceParts.add(result.getPredictedSign());
                currentSign = result.getPredictedSign();
            }
        }
        
        return String.join(" ", sentenceParts);
    }
    
    /**
     * Get model performance info
     */
    public ModelInfo getModelInfo() {
        return new ModelInfo(
            MODEL_PATH,
            INPUT_SIZE,
            OUTPUT_SIZE,
            labels.length,
            "INT8 Quantized",
            interpreter != null ? "Ready" : "Not loaded"
        );
    }
    
    /**
     * Cleanup resources
     */
    public void close() {
        if (interpreter != null) {
            interpreter.close();
            interpreter = null;
        }
        
        if (poseExtractor != null) {
            poseExtractor.close();
            poseExtractor = null;
        }
        
        Log.d(TAG, "✅ ASLRecognizer resources cleaned up");
    }
    
    /**
     * ASL Recognition Result
     */
    public static class ASLResult {
        private final String predictedSign;
        private final float confidence;
        private final String message;
        private final long inferenceTimeMs;
        
        public ASLResult(String predictedSign, float confidence, String message, long inferenceTimeMs) {
            this.predictedSign = predictedSign;
            this.confidence = confidence;
            this.message = message;
            this.inferenceTimeMs = inferenceTimeMs;
        }
        
        // Getters
        public String getPredictedSign() { return predictedSign; }
        public float getConfidence() { return confidence; }
        public String getMessage() { return message; }
        public long getInferenceTimeMs() { return inferenceTimeMs; }
        
        @Override
        public String toString() {
            return String.format("ASLResult{sign='%s', confidence=%.3f, time=%dms}", 
                               predictedSign, confidence, inferenceTimeMs);
        }
    }
    
    /**
     * Model Information
     */
    public static class ModelInfo {
        private final String modelPath;
        private final int inputSize;
        private final int outputSize;
        private final int vocabularySize;
        private final String quantization;
        private final String status;
        
        public ModelInfo(String modelPath, int inputSize, int outputSize, 
                        int vocabularySize, String quantization, String status) {
            this.modelPath = modelPath;
            this.inputSize = inputSize;
            this.outputSize = outputSize;
            this.vocabularySize = vocabularySize;
            this.quantization = quantization;
            this.status = status;
        }
        
        // Getters
        public String getModelPath() { return modelPath; }
        public int getInputSize() { return inputSize; }
        public int getOutputSize() { return outputSize; }
        public int getVocabularySize() { return vocabularySize; }
        public String getQuantization() { return quantization; }
        public String getStatus() { return status; }
    }
}

File: mobile/ios/iOS/Test/Test/ASLRecognizer.swift
import Foundation
import CoreML
import Vision
import UIKit
import AVFoundation
import Combine

/**
 * 🤖 OpenHands ASL Recognition for iOS
 * Core ML implementation with pose-based inference
 */
@available(iOS 14.0, *)
class ASLRecognizer: ObservableObject {
    
    // MARK: - Properties
    
    private var model: MLModel?
    private let confidenceThreshold: Float = 0.7
    private let inputSize = 158  // Pose features
    private let outputSize = 2000  // ASL vocabulary
    
    @Published var isReady = false
    @Published var recognitionHistory: [ASLResult] = []
    
    // MARK: - Initialization
    
    init() {
        setupModel()
    }
    
    /**
     * Setup Core ML model
     */
    private func setupModel() {
        DispatchQueue.global(qos: .userInitiated).async { [weak self] in
            do {
                print("🚀 Loading OpenHands ASL model...")
                
                // Load Core ML model
                let configuration = MLModelConfiguration()
                configuration.computeUnits = .all  // Use GPU if available
                
                // Load the Core ML model directly
                var modelURL: URL?
                
                // Try multiple ways to find the model
                // Method 1: Look for compiled model (.mlmodelc)
                if let compiledURL = Bundle.main.url(forResource: "ASLClassifier", withExtension: "mlmodelc") {
                    modelURL = compiledURL
                    print("✅ Found compiled model: \(compiledURL.path)")
                }
                // Method 2: Look for source model (.mlmodel)  
                else if let sourceURL = Bundle.main.url(forResource: "ASLClassifier", withExtension: "mlmodel") {
                    modelURL = sourceURL
                    print("✅ Found source model: \(sourceURL.path)")
                }
                // Method 3: Debug - list all bundle resources
                else {
                    print("❌ Model not found. Bundle resources:")
                    if let resourcePath = Bundle.main.resourcePath {
                        do {
                            let resources = try FileManager.default.contentsOfDirectory(atPath: resourcePath)
                            for resource in resources.sorted() {
                                print("  - \(resource)")
                            }
                        } catch {
                            print("  Error listing resources: \(error)")
                        }
                    }
                }
                
                guard let url = modelURL else {
                    throw NSError(domain: "ASLRecognizer", code: 1, userInfo: [NSLocalizedDescriptionKey: "ASLClassifier.mlmodel not found in app bundle"])
                }
                self?.model = try MLModel(contentsOf: url, configuration: configuration)
                
                DispatchQueue.main.async {
                    self?.isReady = true
                    print("✅ ASL model loaded successfully")
                }
                
            } catch {
                print("❌ Failed to load ASL model: \(error)")
                DispatchQueue.main.async {
                    self?.isReady = false
                }
            }
        }
    }
    
    // MARK: - Recognition Methods
    
    /**
     * Recognize ASL sign from camera frame
     */
    func recognizeSign(from pixelBuffer: CVPixelBuffer) -> ASLResult {
        let startTime = CFAbsoluteTimeGetCurrent()
        
        guard isReady, let model = model else {
            return ASLResult(
                predictedSign: "",
                confidence: 0.0,
                message: "Model not ready",
                inferenceTimeMs: 0
            )
        }
        
        do {
            // Extract pose features using Vision framework directly
            guard let poseFeatures = extractBasicFeatures(from: pixelBuffer) else {
                let inferenceTime = Int64((CFAbsoluteTimeGetCurrent() - startTime) * 1000)
                return ASLResult(
                    predictedSign: "",
                    confidence: 0.0,
                    message: "No hands detected",
                    inferenceTimeMs: inferenceTime
                )
            }
            
            // Convert to MLMultiArray
            let inputArray = try MLMultiArray(shape: [1, NSNumber(value: inputSize)], dataType: .float32)
            for (index, feature) in poseFeatures.enumerated() {
                if index < inputSize {
                    inputArray[index] = NSNumber(value: feature)
                }
            }
            
            // Run prediction
            let inputDict = ["input": inputArray]
            let prediction = try model.prediction(from: MLDictionaryFeatureProvider(dictionary: inputDict))
            
            // Process output
            guard let outputFeature = prediction.featureValue(for: "output"),
                  let probabilities = outputFeature.multiArrayValue else {
                throw NSError(domain: "ASLRecognizer", code: 2, userInfo: [NSLocalizedDescriptionKey: "Invalid model output"])
            }
            
            var bestIndex = 0
            var bestConfidence: Float = 0.0
            
            // Find highest confidence prediction
            for i in 0..<min(30, probabilities.count) { // Use 30 as our actual class count
                let confidence = probabilities[i].floatValue
                if confidence > bestConfidence {
                    bestConfidence = confidence
                    bestIndex = i
                }
            }
            
            // Create result
            let inferenceTime = Int64((CFAbsoluteTimeGetCurrent() - startTime) * 1000)
            let predictedSign = getSignLabel(for: bestIndex)
            let message = bestConfidence > confidenceThreshold ?
                         "Recognized: \(predictedSign)" :
                         "Low confidence: \(String(format: "%.2f", bestConfidence))"
            
            return ASLResult(
                predictedSign: bestConfidence > confidenceThreshold ? predictedSign : "",
                confidence: bestConfidence,
                message: message,
                inferenceTimeMs: inferenceTime
            )
            
        } catch {
            print("❌ Recognition error: \(error)")
            let inferenceTime = Int64((CFAbsoluteTimeGetCurrent() - startTime) * 1000)
            return ASLResult(
                predictedSign: "",
                confidence: 0.0,
                message: "Error: \(error.localizedDescription)",
                inferenceTimeMs: inferenceTime
            )
        }
    }
    
    /**
     * Recognize ASL sign from UIImage
     */
    func recognizeSign(from image: UIImage) -> ASLResult {
        guard let pixelBuffer = image.toPixelBuffer() else {
            return ASLResult(
                predictedSign: "",
                confidence: 0.0,
                message: "Failed to convert image",
                inferenceTimeMs: 0
            )
        }
        
        return recognizeSign(from: pixelBuffer)
    }
    
    /**
     * Get top N predictions
     */
    func getTopPredictions(from pixelBuffer: CVPixelBuffer, topN: Int = 5) -> [ASLResult] {
        guard isReady, let model = model else { return [] }
        
        do {
            guard let poseFeatures = extractBasicFeatures(from: pixelBuffer) else {
                return []
            }
            
            let inputArray = try MLMultiArray(shape: [1, NSNumber(value: inputSize)], dataType: .float32)
            for (index, feature) in poseFeatures.enumerated() {
                if index < inputSize {
                    inputArray[index] = NSNumber(value: feature)
                }
            }
            
            let inputDict = ["input": inputArray]
            let prediction = try model.prediction(from: MLDictionaryFeatureProvider(dictionary: inputDict))
            
            guard let outputFeature = prediction.featureValue(for: "output"),
                  let probabilities = outputFeature.multiArrayValue else {
                return []
            }
            
            // Create (index, confidence) pairs and sort by confidence
            var indexConfidencePairs: [(Int, Float)] = []
            for i in 0..<min(30, probabilities.count) {
                indexConfidencePairs.append((i, probabilities[i].floatValue))
            }
            
            indexConfidencePairs.sort { $0.1 > $1.1 }  // Sort by confidence descending
            
            // Convert to ASLResults
            let topResults = indexConfidencePairs.prefix(topN).map { index, confidence in
                ASLResult(
                    predictedSign: getSignLabel(for: index),
                    confidence: confidence,
                    message: "Rank prediction",
                    inferenceTimeMs: 0
                )
            }
            
            return topResults
            
        } catch {
            print("❌ Top predictions error: \(error)")
            return []
        }
    }
    
    // MARK: - History Management
    
    /**
     * Add recognition to history
     */
    func addToHistory(_ result: ASLResult) {
        DispatchQueue.main.async { [weak self] in
            self?.recognitionHistory.append(result)
            
            // Keep only recent history (last 50 items)
            if let history = self?.recognitionHistory, history.count > 50 {
                self?.recognitionHistory = Array(history.suffix(50))
            }
        }
    }
    
    /**
     * Clear recognition history
     */
    func clearHistory() {
        DispatchQueue.main.async { [weak self] in
            self?.recognitionHistory.removeAll()
        }
    }
    
    /**
     * Build sentence from recent recognitions
     */
    func buildSentence() -> String {
        let recentResults = recognitionHistory.suffix(20)  // Last 20 recognitions
        var sentenceParts: [String] = []
        var currentSign: String?
        
        // Group consecutive similar signs
        for result in recentResults {
            if result.confidence > confidenceThreshold &&
               !result.predictedSign.isEmpty &&
               result.predictedSign != currentSign {
                
                sentenceParts.append(result.predictedSign)
                currentSign = result.predictedSign
            }
        }
        
        return sentenceParts.joined(separator: " ")
    }
    
    // MARK: - Feature Extraction
    
    /**
     * Extract basic features from pixel buffer for recognition
     */
    private func extractBasicFeatures(from pixelBuffer: CVPixelBuffer) -> [Float]? {
        // Simple feature extraction using Vision framework
        let request = VNDetectHumanHandPoseRequest()
        let handler = VNImageRequestHandler(cvPixelBuffer: pixelBuffer, options: [:])
        
        do {
            try handler.perform([request])
            
            guard let observations = request.results, !observations.isEmpty else {
                return nil
            }
            
            var features: [Float] = []
            
            // Extract hand landmarks (simplified)
            for observation in observations.prefix(2) { // Max 2 hands
                // Get key landmarks
                let landmarkKeys: [VNHumanHandPoseObservation.JointName] = [
                    .wrist, .thumbTip, .indexTip, .middleTip, .ringTip, .littleTip
                ]
                
                for key in landmarkKeys {
                    if let point = try? observation.recognizedPoint(key) {
                        features.append(Float(point.location.x))
                        features.append(Float(point.location.y))
                        features.append(Float(point.confidence))
                    } else {
                        features.append(contentsOf: [0.0, 0.0, 0.0])
                    }
                }
            }
            
            // Pad features to expected input size
            while features.count < inputSize {
                features.append(0.0)
            }
            
            return Array(features.prefix(inputSize))
            
        } catch {
            print("❌ Feature extraction failed: \(error)")
            return nil
        }
    }
    
    // MARK: - Utilities
    
    /**
     * Get sign label for index
     */
    private func getSignLabel(for index: Int) -> String {
        // Load labels from bundle or use default mapping
        if let labelsPath = Bundle.main.path(forResource: "asl_labels", ofType: "txt"),
           let labelsContent = try? String(contentsOfFile: labelsPath),
           !labelsContent.isEmpty {
            
            let labels = labelsContent.components(separatedBy: .newlines)
            return index < labels.count ? labels[index] : "UNKNOWN_\(index)"
        }
        
        // Fallback to generic labels
        return "SIGN_\(index)"
    }
    
    /**
     * Get model performance information
     */
    func getModelInfo() -> ModelInfo {
        return ModelInfo(
            modelPath: "ASLClassifier.mlmodel",
            inputSize: inputSize,
            outputSize: outputSize,
            vocabularySize: outputSize,
            quantization: "Core ML Optimized",
            status: isReady ? "Ready" : "Loading",
            device: getComputeDevice()
        )
    }
    
    private func getComputeDevice() -> String {
        guard let model = model else { return "Unknown" }
        
        // Try to determine compute device
        if #available(iOS 14.0, *) {
            // Check model configuration
            return "Neural Engine/GPU"
        } else {
            return "CPU"
        }
    }
}


// MARK: - UIImage Extension

extension UIImage {
    /**
     * Convert UIImage to CVPixelBuffer
     */
    func toPixelBuffer() -> CVPixelBuffer? {
        let width = Int(self.size.width)
        let height = Int(self.size.height)
        
        let attrs = [
            kCVPixelBufferCGImageCompatibilityKey: kCFBooleanTrue,
            kCVPixelBufferCGBitmapContextCompatibilityKey: kCFBooleanTrue
        ] as CFDictionary
        
        var pixelBuffer: CVPixelBuffer?
        let status = CVPixelBufferCreate(
            kCFAllocatorDefault,
            width,
            height,
            kCVPixelFormatType_32ARGB,
            attrs,
            &pixelBuffer
        )
        
        guard status == kCVReturnSuccess, let buffer = pixelBuffer else {
            return nil
        }
        
        CVPixelBufferLockBaseAddress(buffer, CVPixelBufferLockFlags(rawValue: 0))
        let pixelData = CVPixelBufferGetBaseAddress(buffer)
        
        let colorSpace = CGColorSpaceCreateDeviceRGB()
        guard let context = CGContext(
            data: pixelData,
            width: width,
            height: height,
            bitsPerComponent: 8,
            bytesPerRow: CVPixelBufferGetBytesPerRow(buffer),
            space: colorSpace,
            bitmapInfo: CGImageAlphaInfo.noneSkipFirst.rawValue
        ) else {
            CVPixelBufferUnlockBaseAddress(buffer, CVPixelBufferLockFlags(rawValue: 0))
            return nil
        }
        
        context.translateBy(x: 0, y: CGFloat(height))
        context.scaleBy(x: 1, y: -1)
        
        UIGraphicsPushContext(context)
        self.draw(in: CGRect(x: 0, y: 0, width: width, height: height))
        UIGraphicsPopContext()
        
        CVPixelBufferUnlockBaseAddress(buffer, CVPixelBufferLockFlags(rawValue: 0))
        
        return buffer
    }
}
